src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs

```diff
@@ -0,0 +1,196 @@
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    /// <summary>
+    /// Discourages anyone instantiating their own JoinableTaskContext within VS.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class VSSDK006CheckServicesExistAnalyzer : DiagnosticAnalyzer
+    {
+        /// <summary>
+        /// The value to use for <see cref="DiagnosticDescriptor.Id"/> in generated diagnostics.
+        /// </summary>
+        public const string Id = "VSSDK006";
+
+        /// <summary>
+        /// A reusable descriptor for diagnostics produced by this analyzer.
+        /// </summary>
+        internal static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
+            id: Id,
+            title: "Check services exist",
+            messageFormat: "Check whether the result of GetService calls is null.",
+            helpLinkUri: Utils.GetHelpLink(Id),
+            category: "Reliability",
+            defaultSeverity: DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterCompilationStartAction(start =>
+            {
+                var state = new CompilationState(
+                    Flatten(
+                        start.Compilation.GetTypeByMetadataName(Types.Package.FullName)?.GetMembers(Types.Package.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.AsyncPackage.FullName)?.GetMembers(Types.AsyncPackage.GetServiceAsync),
+                        start.Compilation.GetTypeByMetadataName(Types.AsyncPackage.FullName)?.GetMembers(Types.Package.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.ServiceProvider.FullName).GetMembers(Types.ServiceProvider.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.IServiceProvider.FullName).GetMembers(Types.IServiceProvider.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.IAsyncServiceProvider.FullName).GetMembers(Types.IAsyncServiceProvider.GetServiceAsync)),
+                    Flatten(
+                        start.Compilation.GetTypeByMetadataName(Types.Assumes.FullName).GetMembers(Types.Assumes.Present)));
+                if (state.ShouldAnalyze)
+                {
+                    start.RegisterSyntaxNodeAction(state.AnalyzeInvocationExpression, SyntaxKind.InvocationExpression);
```

Comment:
Did you consider `RegisterOperationAction` with [OperationKind.Invocation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/OperationKind.cs,57)?

Reply:
Briefly, yes. Operations is a roslyn feature that isn't always turned on, as I understand it. It may have been stabilized and turned on by default at some point in Roslyn's timeline but I believe in Dev14 (which our analyzers typically target) I think operations were off, so we don't depend on them since that would reduce our audience.

Is there some significant benefit we should be considering?